### PR TITLE
KAFKA-7439: Replace EasyMock and PowerMock with Mockito in clients module

### PR DESCRIPTION
**Pull request by ijuma**
Development of EasyMock and PowerMock has stagnated while Mockito
continues to be actively developed. With the new Java release cadence,
it's a problem to depend on libraries that do bytecode manipulation
and are not actively maintained. In addition, Mockito is also
easier to use.

While updating the tests, I attempted to go from failing test to
passing test. In cases where the updated test passed on the first
attempt, I artificially broke it to ensure the test was still doing its
job.

I included a few improvements that were helpful while making these
changes:

1. Better exception if there are no nodes in `leastLoadedNodes`
2. Always close the producer in `KafkaProducerTest`
3. requestsInFlight producer metric should not hold a reference to
`Sender`

Finally, `Metadata` is no longer final so that we don't need
`PowerMock` to mock it. It's an internal class, so it's OK.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 1]
```diff
@@ -50,7 +50,7 @@
  * is removed from the metadata refresh set after an update. Consumers disable topic expiry since they explicitly
  * manage topics while producers rely on topic expiry to limit the refresh set.
  */
-public final class Metadata implements Closeable {
+public class Metadata implements Closeable {
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:53]: Is this change needed because of mocking?

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 2]
```diff
@@ -595,6 +595,8 @@ public void close() {
     @Override
     public Node leastLoadedNode(long now) {
         List<Node> nodes = this.metadataUpdater.fetchNodes();
+        if (nodes.isEmpty())
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:598]: Does `nodes.isEmpty()` always mean that an empty cluster is initialized? Is throwing this exception differentiates this case from the one where the `leastLoadedNode` method wasn't able to select the least loaded from a set of existing (but maybe disconnected) nodes?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java [hunk 3]
```diff
@@ -241,36 +229,25 @@ public void testCleanupMemoryAvailabilityWaiterOnInterruption() throws Exception
...
-        expect(mockedSensor.add(new Meter(TimeUnit.NANOSECONDS, rateMetricName, totalMetricName))).andReturn(true);
-
-        replay(mockedMetrics, mockedSensor, metricName);
-
-        BufferPool bufferPool = new BufferPool(2, 1, mockedMetrics, time,  metricGroup);
         bufferPool.allocate(1, 0);
         try {
             bufferPool.allocate(2, 1000);
-            assertTrue("Expected oom.", false);
+            fail("Expected oom.");
```
Comment (viktorsomogyi) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java:241]: nit: I usually prefer using the [ExpectedException](https://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html) feature of junit as it would describe the test failure better (what was the actual exception vs the expected) and a bit more concise.

**Review by viktorsomogyi — COMMENTED** [review]
Added a few comments but it generally looks good to me. Also ran your patch on our Jenkins and it passed for the first time, no flakes seen.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 1 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java:53]: Yes.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java (hunk 3 above)
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java:241]: Yeah, but JUnit 5 has a better way via lambdas, so I'd rather change to that (I have an in-progress PR that does that, but this one should be merged first).

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java (hunk 2 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:598]: `int offset = this.randOffset.nextInt(nodes.size());` will fail if `nodes.isEmpty()` is true, but it's a more cryptic exception. So, we simply get a more useful error now.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 1 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/Metadata.java:53]: I updated the PR description to mention this now.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java (hunk 2 above)
Comment (viktorsomogyi) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:598]: Ok, thanks for explaining.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java (hunk 3 above)
Comment (viktorsomogyi) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/BufferPoolTest.java:241]: Cool, looking forward to it :)

**Comment by ijuma** [comment]
@lindong28 Would you have the time to review this?

**Comment by ijuma** [comment]
Mockito 2.23.0 with full support for Java 11 was released, so I updated the dep and rebased (no conflicts thankfully).

**Comment by ijuma** [comment]
@lindong28, just checking, will you have time to review this PR? If you're busy, I can try someone else. Thanks!

**Comment by lindong28** [comment]
@ijuma Yes! I am planning to look through this tonight :D

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 4]
```diff
@@ -353,134 +360,145 @@ public void testMetadataFetchOnStaleMetadata() throws Exception {
...
         final String topic = "topic";
-
-        Thread t = new Thread(() -> {
-            long startTimeMs = System.currentTimeMillis();
-            for (int i = 0; i < 10; i++) {
-                while (!metadata.updateRequested() && System.currentTimeMillis() - startTimeMs < 1000)
-                    Thread.yield();
-                metadata.update(Cluster.empty(), Collections.singleton(topic), time.milliseconds());
-                time.sleep(60 * 1000L);
+        try (KafkaProducer<String, String> producer = new KafkaProducer<>(config, null, null, metadata,
```
Comment (lindong28) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:433]: Would it be more realistic and consistent with old code to do `new KafkaProducer<>(config, new StringSerializer(), new StringSerializer(), metadata, null, null, time))` here? The new code at line 488 seems to be consistent with the old code.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 5]
```diff
@@ -522,40 +541,38 @@ public void testMetricConfigRecordingLevel() {
...
-        ProducerInterceptors<String, String> interceptors = PowerMock.createMock(ProducerInterceptors.class);
-        EasyMock.expect(interceptors.onSend(record)).andReturn(record);
-        interceptors.onSendError(EasyMock.eq(record), EasyMock.notNull(), EasyMock.notNull());
-        EasyMock.expectLastCall();
-        MemberModifier.field(KafkaProducer.class, "interceptors").set(producer, interceptors);
-
-        PowerMock.replay(metadata);
-        EasyMock.replay(interceptors);
+                ProducerInterceptors<String, String> interceptors = mock(ProducerInterceptors.class);
+        KafkaProducer<String, String> producer = new KafkaProducer<>(config, null, null,
```
Comment (lindong28) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:565]: Same as above.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 6]
```diff
@@ -589,7 +606,7 @@ public void testInitTransactionTimeout() {
 
         try (Producer<String, String> producer = new KafkaProducer<>(
                 new ProducerConfig(ProducerConfig.addSerializerToConfig(props, new StringSerializer(), new StringSerializer())),
-                new StringSerializer(), new StringSerializer(), metadata, client)) {
+                null, null, metadata, client, null, time)) {
```
Comment (lindong28) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:609]: Same as above.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 7]
```diff
@@ -614,7 +631,7 @@ public void testOnlyCanExecuteCloseAfterInitTransactionsTimeout() {
 
         Producer<String, String> producer = new KafkaProducer<>(
                 new ProducerConfig(ProducerConfig.addSerializerToConfig(props, new StringSerializer(), new StringSerializer())),
-                new StringSerializer(), new StringSerializer(), metadata, client);
+                null, null, metadata, client, null, time);
```
Comment (lindong28) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:634]: Same as above.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 8]
```diff
@@ -647,9 +663,9 @@ public void testSendToInvalidTopic() throws Exception {
 
         Producer<String, String> producer = new KafkaProducer<>(new ProducerConfig(
                 ProducerConfig.addSerializerToConfig(props, new StringSerializer(), new StringSerializer())),
-                new StringSerializer(), new StringSerializer(), metadata, client);
+                null, null, metadata, client, null, time);
```
Comment (lindong28) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:666]: Same as above.

File: clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java [hunk 9]
```diff
@@ -295,25 +290,24 @@ public void testTruncateNotCalledIfSizeIsBiggerThanTargetSize() throws IOExcepti
...
     /**
      * see #testTruncateNotCalledIfSizeIsSameAsTargetSize
      */
     @Test
     public void testTruncateIfSizeIsDifferentToTargetSize() throws IOException {
-        FileChannel channelMock = EasyMock.createMock(FileChannel.class);
+        FileChannel channelMock = mock(FileChannel.class);
 
-        EasyMock.expect(channelMock.size()).andReturn(42L).atLeastOnce();
-        EasyMock.expect(channelMock.position(42L)).andReturn(null).once();
```
Comment (lindong28) [inline clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java:309]: Do we still need to verify that `channelMock.position` is called once?

Given that the old test works, `channelMock.position(...)` should be invoked. If we don't do `when(channelMock.position(42L)).thenReturn(null)`, what would be the return value if `channelMock.position(...)` is called?

It seems that channelMock.position() is called in the old code but not the new code. If so I am wondering which part changes this behavior.

**Review by lindong28 — COMMENTED** [review]
Thanks much for the patch. Looks great over all. Left some minor comments.

File: clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java (hunk 9 above)
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java:309]: I removed that because it didn't seem relevant to this test. Mockito handles the `position` call correctly without us doing anything. With EasyMock, if not using a `Nice` mock, we have to specify every method called.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 4 above)
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:433]: It's a good question. The problem is that we have to set the serializers in `ProducerConfig`. Once we do that, we can just pass `null` for both of these to make it obvious that they have been set already. Otherwise, you can end up with inconsistent serializers by updating them in one place and forgetting to update them elsewhere.

This is ugly though and I wasn't too happy with it, so I changed the producer a bit to simplify this. Let me know what you think.

File: clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java (hunk 9 above)
Comment (lindong28) [reply clients/src/test/java/org/apache/kafka/common/record/FileRecordsTest.java:309]: I see. Sounds good.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 4 above)
Comment (lindong28) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:433]: Previously the test code used constructor `KafkaProducer(Map<String, Object> configs, Serializer<K> keySerializer, Serializer<V> valueSerializer)`, which translates to `KafkaProducer(new ProducerConfig(ProducerConfig.addSerializerToConfig(configs, keySerializer, valueSerializer)),keySerializer, valueSerializer, null, null, null, Time.SYSTEM)` where keySerializer and valueSerializer are not null. I was thinking that the goal this patch is to mimic the same behavior.

Now I look through the `KafkaProducer` code again, I realized there is constructor `KafkaProducer(final Map<String, Object> configs)` which translates to `KafkaProducer(new ProducerConfig(configs), null, null, null, null, null, Time.SYSTEM)`. It is OK for the test code to switch to this constructor after explicitly using `addSerializerToConfig(...)`.

Thanks for the update. It looks like a lot of test failed after the most recent commit. Maybe it is simpler to just revert the recent change and focus on the migration to mockito in this PR.

Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:433]: Yes, I'll revert and do a separate PR for the simplification while making the whole test suite pass.

Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:433]: Btw, I left line 488 as it is because there's what I'd call a bug in `ProducerConfig.addSerializerToConfig` where it doesn't use the actual instance, just the class name. However, it's a public method so not sure if people are relying on it. This will go away in the PR that has the commit I just reverted so I think we can merge as it is.

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
The test failure is not related. @lindong28, do you have any additional concerns?

**Review by lindong28 — APPROVED** [review]
Thanks for the patch! LGTM. Merged to trunk.

**Comment by ijuma** [comment]
@lindong28 Since this is mostly test changes, would you be OK with cherry-picking to 2.1? Along with the small https://github.com/apache/kafka/pull/5771, we can claim Java 11 support in Kafka 2.1.0.

**Comment by lindong28** [comment]
@ijuma Certainly. Since it is useful to claim Java 11 support, I will cherry-pick this to 2.1 branch.

**Comment by lindong28** [comment]
This PR has been cherry-picked to 2.1 branch.

**Comment by ijuma** [comment]
Thanks!
